### PR TITLE
fix: use different separator to flatten dict

### PR DESCRIPTION
**Pull request by Captainia**
*Issue #, if available:*

*Description of changes:* We need a different separator when flatten the dict and merge. This is because `ml.p4d.24xlarge` got treated as nested dict with the original separator `.`.

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/unit/test_utils.py [hunk 1]
```diff
@@ -1837,7 +1837,7 @@ def test_flatten_dict_no_nested(self):
 
     def test_flatten_dict_with_various_types(self):
         nested_dict = {"a": [1, 2, 3], "b": {"x": None, "y": {"p": [], "q": ""}}, "c": 9}
-        flattened_dict = {"a": [1, 2, 3], "b.x": None, "b.y.p": [], "b.y.q": "", "c": 9}
+        flattened_dict = {"a": [1, 2, 3], "b^x": None, "b^y^p": [], "b^y^q": "", "c": 9}
```
Comment (evakravi) [inline tests/unit/test_utils.py:1840]: Can you add a unit test that combines instance-specific metadata, to re-create the situation Kyle had tested?

Comment (Captainia) [reply tests/unit/test_utils.py:1840]: I'll address this in my next PR which the tools will be replaced

**Review by makungaj1 — APPROVED** [review]

**Comment by Aditi2424** [comment]
@Captainia is this going to be backward compatible?

**Comment by Captainia** [comment]
> @Captainia is this going to be backward compatible?
Yes this is a new util that we added for the feature branch, this is not something used across other features in mainline.

**Review by Aditi2424 — APPROVED** [review]

**Pull request merged**